pkg/service/one2onerestore/worker_tgc.go

```diff
@@ -0,0 +1,181 @@
+func (w *worker) setTombstoneGCModeRepair(ctx context.Context, workload []hostWorkload, keyspaceFilter []string) error {
+	w.awaitSchemaAgreement(ctx, w.clusterSession)
+
+	type backupTable struct {
+		keyspace, table string
+	}
+
+	tablesToRestore := map[backupTable]struct{}{}
+
+	for _, wl := range workload {
+		err := wl.manifestContent.ForEachIndexIter(keyspaceFilter, func(fm backupspec.FilesMeta) {
+			tablesToRestore[backupTable{keyspace: fm.Keyspace, table: fm.Table}] = struct{}{}
+		})
+		if err != nil {
+			return errors.Wrap(err, "manifest content files")
+		}
+	}
+
+	for table := range tablesToRestore {
+		mode, err := w.getTableTombstoneGCMode(table.keyspace, table.table)
+		if err != nil {
+			return errors.Wrap(err, "get tombstone_gc mode")
+		}
+		// No need to change tombstone gc mode.
+		if mode == modeDisabled || mode == modeImmediate || mode == modeRepair {
+			w.logger.Info(ctx, "Skipping set tombstone_gc mode", "table", table, "mode", mode)
+			continue
+		}
+		if err := w.setTableTombstoneGCMode(ctx, table.keyspace, table.table, modeRepair); err != nil {
+			return errors.Wrap(err, "set tombstone_gc mode repair")
+		}
+	}
+
+	return nil
+}
+
+// getTableTombstoneGCMode returns table's tombstone_gc mode.
+func (w *worker) getTableTombstoneGCMode(keyspace, table string) (tombstoneGCMode, error) {
+	var ext map[string]string
+	q := qb.Select("system_schema.tables").
+		Columns("extensions").
+		Where(qb.Eq("keyspace_name"), qb.Eq("table_name")).
+		Query(w.clusterSession).
+		Bind(keyspace, table)
+
+	defer q.Release()
+	err := q.Scan(&ext)
+	if err != nil {
+		return "", err
+	}
+
+	// Timeout (just using gc_grace_seconds) is the default mode
+	mode, ok := ext["tombstone_gc"]
+	if !ok {
+		return modeTimeout, nil
+	}
+
+	allModes := []tombstoneGCMode{modeDisabled, modeTimeout, modeRepair, modeImmediate}
+	for _, m := range allModes {
+		if strings.Contains(mode, string(m)) {
+			return m, nil
+		}
+	}
+	return "", errors.Errorf("unrecognized tombstone_gc mode: %s", mode)
+}
+
+// setTableTombstoneGCMode alters 'tombstone_gc' mode.
+func (w *worker) setTableTombstoneGCMode(ctx context.Context, keyspace, table string, mode tombstoneGCMode) error {
+	w.logger.Info(ctx, "Alter table's tombstone_gc mode",
+		"keyspace", keyspace,
+		"table", table,
+	)
+
+	op := func() error {
+		return w.clusterSession.ExecStmt(alterTableTombstoneGCStmt(keyspace, table, mode))
+	}
+
+	notify := func(err error, wait time.Duration) {
+		w.logger.Info(ctx, "Altering table's tombstone_gc mode failed",
+			"keyspace", keyspace,
+			"table", table,
+			"error", err,
+			"wait", wait,
+		)
+	}
+
+	return alterSchemaRetryWrapper(ctx, op, notify)
+}
+
+func alterTableTombstoneGCStmt(keyspace, table string, mode tombstoneGCMode) string {
+	return fmt.Sprintf(`ALTER TABLE %q.%q WITH tombstone_gc = {'mode': '%s'}`, keyspace, table, mode)
+}
+
+func (w *worker) awaitSchemaAgreement(ctx context.Context, clusterSession gocqlx.Session) {
```

Comment:
I don't think that it's needed here. This is copied from the regular restore, which was copied from the backup, where it made sense to await schema agreement before backing up schema sstables. Here we are even using some additional wrappers over schema altering methods in case they take more time because of many recent schema changes.

Reply:
Not to mention that this implementation probably requires single host session to work properly, as the two queries might be routed to different hosts.